readthedocs/donate/signals.py

```diff
@@ -1,9 +1,118 @@
+import random
+
 from django.dispatch import receiver
 from django.conf import settings
 from django.core.cache import cache
 
 from readthedocs.restapi.signals import footer_response
-from readthedocs.donate.models import SupporterPromo, VIEWS, CLICKS, OFFERS
+from readthedocs.donate.models import SupporterPromo, VIEWS, CLICKS, OFFERS, INCLUDE, EXCLUDE
+from readthedocs.donate.utils import get_ad_day
+
+
+PROMO_GEO_PATH = getattr(settings, 'PROMO_GEO_PATH', None)
+
+if PROMO_GEO_PATH:
+    import geoip2.database  # noqa
+    geo_reader = geoip2.database.Reader(PROMO_GEO_PATH)
+
+
+def show_to_geo(promo, country_code):
+    # Remove promo's that exclude this country.
+    for filter in promo.geo_filters.all():
+        if filter.filter_type == INCLUDE:
+            if country_code in filter.codes:
+                continue
+            else:
+                return False
+        if filter.filter_type == EXCLUDE:
+            if country_code in filter.codes:
+                return False
```

Comment:
This seems like it should be simple to push on to a queryset. The added complication that I see is that the exclude filters assume an implicit include filter if there are no other filters. If not a straight chain of filters, simply iterating over the filters to build the queryset is also an option.

I'm still not convinced that we don't want to model this as a straight many to many between promo and countries, not as a filter mechanism. It seems adding countries as an include/exclude is strictly a management function, not a modeling one.